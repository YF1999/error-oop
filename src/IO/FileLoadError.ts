import { IOError } from './IOError';

export class FileLoadError extends IOError {
    private _fileName?: string;

    public constructor();
    public constructor(message: string);
    public constructor(message: string, innerError: Error);
    public constructor(message: string, fileName: string);
    public constructor(message: string, fileName: string, innerError: Error);

    public constructor(message?: string, arg1?: string | Error, arg2?: Error) {
        // message?
        if (message === undefined) {
            super();
        }

        // message + innerError?/fileName?
        else if (arg1 === undefined) {
            super(message);
        }

        // message + innerError
        else if (typeof arg1 !== 'string') {
            super(message, arg1);
        }

        // message + fileName + innerError?
        else if (arg2 === undefined) {
            super(message);
            this._fileName = arg1;
        }

        // message + fileName + innerError
        else {
            super(message, arg2);
            this._fileName = arg1;
        }
    }

    get fileName() {
        return this._fileName;
    }
}
