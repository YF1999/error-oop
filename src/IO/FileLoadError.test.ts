import { FileLoadError } from './FileLoadError';

describe('Test `FileLoadError`', () => {
    const name = 'FileLoadError';
    const message = 'error';
    const fileName = 'text.txt';

    function getErrors() {
        const err1 = new FileLoadError();
        const err2 = new FileLoadError(message);
        const err3 = new FileLoadError(message, err2);
        const err4 = new FileLoadError(message, fileName, err3);
        const err5 = new FileLoadError({ message, innerError: err4 });
        const err6 = new FileLoadError({ message, innerError: err5, fileName });

        return [err1, err2, err3, err4, err5, err6];
    }

    test('property: error name', () => {
        getErrors().map((err) => expect(err.name).toBe(name));
    });

    test('property: message', () => {
        const [err1, ...errs] = getErrors();
        expect(err1.message).toBe('');
        errs.map((err) => expect(err.message).toBe(message));
    });

    test('property: inner error', () => {
        const [err1, err2, err3, err4, err5, err6] = getErrors();
        expect(err1.innerError).toBeNull();
        expect(err2.innerError).toBeNull();
        expect(err3.innerError).toBe(err2);
        expect(err4.innerError).toBe(err3);
        expect(err5.innerError).toBe(err4);
        expect(err6.innerError).toBe(err5);
    });

    test('property: file name', () => {
        const [err1, err2, err3, err4, err5, err6] = getErrors();
        expect(err1.fileName).toBeNull();
        expect(err2.fileName).toBeNull();
        expect(err3.fileName).toBeNull();
        expect(err4.fileName).toBe(fileName);
        expect(err5.fileName).toBeNull();
        expect(err6.fileName).toBe(fileName);
    });

    test('empty serialization', () => {
        getErrors().map((err) => expect(JSON.stringify(err)).toBe('{}'));
    });

    test('no enumerable fields', () => {
        getErrors().map((err) => expect(Object.keys(err).length).toBe(0));
    });

    test('own keys', () => {
        const [err1, ...errs] = getErrors();
        expect(Reflect.ownKeys(err1)).toStrictEqual(['stack', '_fileName']);
        errs.map((err) => expect(Reflect.ownKeys(err)).toStrictEqual(['stack', 'message', '_fileName']));
    });
});
