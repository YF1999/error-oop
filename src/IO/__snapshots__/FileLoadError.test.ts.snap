// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test \`FileLoadError\` own keys 1`] = `
Array [
  "stack",
  "_fileName",
]
`;

exports[`Test \`FileLoadError\` own keys 2`] = `
Array [
  "stack",
  "message",
  "_fileName",
]
`;

exports[`Test \`FileLoadError\` own keys 3`] = `
Array [
  "stack",
  "message",
  "_fileName",
]
`;

exports[`Test \`FileLoadError\` own keys 4`] = `
Array [
  "stack",
  "message",
  "_fileName",
]
`;

exports[`Test \`FileLoadError\` own keys 5`] = `
Array [
  "stack",
  "message",
  "_fileName",
]
`;

exports[`Test \`FileLoadError\` own keys 6`] = `
Array [
  "stack",
  "message",
  "_fileName",
]
`;
